server/gateway.go

```diff
@@ -1849,3 +2108,121 @@ func (c *client) processInboundGatewayMsg(msg []byte) {
+// subscriptions and we should now not send unless we know
+// there is explicit interest.
+// <Invoked from outbound connection's readLoop>
+func (c *client) gatewayAllSubsReceiveComplete(info *Info) {
+	account := getAccountFromGatewayCommand(c, info, "complete")
+	if account == "" {
+		return
+	}
+	// Done receiving all subs from remote. Set the `ni`
+	// map to nil so that gatewayInterest() no longer
+	// uses it.
+	ei, _ := c.gw.outsim.Load(string(account))
+	if ei != nil {
+		e := ei.(*outsie)
+		// Needs locking here since `ni` is checked by
+		// many go-routines calling gatewayInterest()
+		e.Lock()
+		e.ni = nil
+		e.Unlock()
+	}
+}
+
+// small helper to get the account name from the INFO command.
+func getAccountFromGatewayCommand(c *client, info *Info, cmd string) string {
+	if info.GatewayCmdPayload == nil {
+		c.sendErrAndErr(fmt.Sprintf("Account absent from receive-all-subscriptions-%s command", cmd))
+		c.closeConnection(ProtocolViolation)
+		return ""
+	}
+	return string(info.GatewayCmdPayload)
+}
+
+// Switch to send-all-subs mode for the given gateway and account.
+// This is invoked when processing an inbound message and we
+// reach a point where we had to send a lot of RS- for this
+// account. We will send an INFO protocol to indicate that we
+// start sending all our subs (for this account), followed by
+// all subs (RS+) and finally an INFO to indicate the end of it.
+// The remote will then send messages only if it finds explicit
+// interest in the sublist created based on all RS+ that we just
+// sent.
+// The client's lock is held on entry.
+// <Invoked from inbound connection's readLoop>
+func (c *client) gatewaySwitchAccountToSendAllSubs(e *insie) {
+	// Set this map to nil so that the no-interest is
+	// no longer checked.
+	e.ni = nil
+	// Capture this since we are passing it to a go-routine.
+	account := string(c.pa.account)
+	s := c.srv
+
+	// Function that will create an INFO protocol
+	// and set proper command.
+	sendCmd := func(cmd byte, useLock bool) {
+		// Use bare server info and simply set the
+		// gateway name and command
+		info := Info{
+			Gateway:           s.getGatewayName(),
+			GatewayCmd:        cmd,
+			GatewayCmdPayload: []byte(account),
+		}
+		b, _ := json.Marshal(&info)
+		infoJSON := []byte(fmt.Sprintf(InfoProto, b))
+		if useLock {
+			c.mu.Lock()
+		}
+		c.sendProto(infoJSON, true)
+		if useLock {
+			c.mu.Unlock()
+		}
+	}
+	// Send the start command. When remote receives this,
+	// it may continue to send optimistic messages, but
+	// it will start to register RS+/RS- in sublist instead
+	// of noInterest map.
+	sendCmd(gatewayCmdAllSubsStart, false)
+
+	// Execute this in separate go-routine as to not block
+	// the readLoop (which may cause the otherside to close
+	// the connection due to slow consumer)
+	s.startGoRoutine(func() {
+		defer s.grWG.Done()
+
+		s.sendAccountSubsToGateway(c, []byte(account))
+		// Send the complete command. When the remote receives
```

Comment:
Can something proto wise get in between these? Does it matter?

Reply:
It can (like a PING, etc..) and it does not matter. Then "end" command is simply for the remote to optimize and clear the map that it uses to know if it should look at no-interest or not. Until it gets this message it is a bit of a transition phase, but nothing will break at all.

Reply:
So an extra RS- or RS+ won't matter?

Reply:
When we do the actual send, there is a lock that prevents subs/unsubs from happening (state stored in the pasi map). So we are sure to prevent missing some.
Also note that earlier, when deciding to switch to this mode, we set a map to nil which indicate to the code that checks if it should send that it should not attempt (this map is set to nil with a common lock). Finally, when all has been sent, we set a flag ("sas", which I will rename) to instruct that now when a sub is created, it should be sent, of course we do still send once per subject/queue, etc.. Finally, the receiving side of these protos is made idempotent in that if by any chance we were to send the same proto twice, it would not matter.

Reply:
and by the way, we only send plain subs because we always send QSubs.